components/brave_wallet/browser/cardano/cardano_knapsack_solver.cc

```diff
@@ -0,0 +1,177 @@
+    // yet not picked groups(starting from largest).
+    for (int pass = 0; pass < 2; ++pass) {
+      if (has_valid_transaction_for_iteration) {
+        DCHECK_EQ(pass, 1);
+        break;
+      }
+
+      for (uint32_t input_index = 0; input_index < inputs_.size();
+           ++input_index) {
+        bool pick_group =
+            pass == 0 ? base::RandInt(0, 1) : !picked_inputs[input_index];
+        if (!pick_group) {
+          continue;
+        }
+
+        CardanoTransaction next_transaction = cur_transaction;
+        next_transaction.AddInput(inputs_[input_index]);
+        next_transaction.AddWitness(
+            CardanoTransaction::TxWitness::DummyTxWitness());
+
+        // Minimum fee required for this transaction to be accepted.
+        // Depends on transaction's size and current fee rate.
+        uint64_t min_fee = ApplyFeeRate(
+            min_fee_coefficient_, min_fee_constant_,
+            CardanoSerializer::CalcTransactionSize(next_transaction));
+
+        // Move everything except `min_fee` to change output(if any). Throw
+        // away possible transaction if resulting change amount is less than
+        // dust threshold.
+        if (auto change_amount =
+                next_transaction.MoveSurplusFeeToChangeOutput(min_fee)) {
+          if (change_amount < dust_output_threshold) {
+            continue;
+          }
+        }
+
+        if (next_transaction.AmountsAreValid(min_fee)) {
+          has_valid_transaction_for_iteration = true;
+          if (!current_best_solution ||
+              current_best_solution->EffectiveFeeAmount() >
+                  next_transaction.EffectiveFeeAmount()) {
+            current_best_solution = next_transaction;
+          }
+        } else {
+          picked_inputs[input_index] = true;
+          cur_transaction = std::move(next_transaction);
+        }
+      }
+    }
+  }
+}
+
+base::expected<CardanoTransaction, std::string> CardanoKnapsackSolver::Solve() {
+  DCHECK_EQ(base_transaction_.inputs().size(), 0u);
+  DCHECK(base_transaction_.TargetOutput());
+  DCHECK(base_transaction_.ChangeOutput());
+  DCHECK(!base_transaction_.sending_max_amount());
+
+  std::optional<CardanoTransaction> best_solution;
+
+  // Try to find the best transaction with a change output which receives a
+  // fee surplus.
+  SolveForTransaction(base_transaction_, best_solution);
+
+  // Drop the change output from the transaction and try to find the best
+  // transaction again. Might find a transaction with a slightly higher fee
+  // but still less than the cost of having a change output.
+  auto no_change_transaction = base_transaction_;
+  no_change_transaction.ClearChangeOutput();
+  SolveForTransaction(no_change_transaction, best_solution);
+
+  if (!best_solution) {
+    return base::unexpected(
+        l10n_util::GetStringUTF8(IDS_BRAVE_WALLET_INSUFFICIENT_BALANCE));
+  }
+
+  best_solution->SetWitnesses({});
```

Comment:
why is this being set to an empty list here?

Reply:
added comment